### PR TITLE
chore: E2E Test Support

### LINKED ISSUE #99: Create end-to-end test suite
Leverage the [test-harness](https://github.com/open-feature/test-harness) and [evaluation.feature](https://github.com/open-feature/test-harness/blob/main/features/evaluation.feature) gherkin suite to create a set of integration tests for the Python SDK. See equivalent examples in [Java](https://github.com/open-feature/java-sdk/tree/main/src/test/java/dev/openfeature/sdk/e2e) and [JS](https://github.com/open-feature/js-sdk/tree/main/packages/server/e2e). Though we should call these "end-to-end" not integration tests, since that's what they are.

The suite should start the test harness container and run it in CI (again, see references in [Java](https://github.com/open-feature/java-sdk/blob/main/.github/workflows/pullrequest.yml) and [JS](https://github.com/open-feature/js-sdk/blob/main/.github/workflows/pr-checks.yaml)).

NOTE: we'll need the [python flagd provider](https://github.com/open-feature/python-sdk-contrib/pull/3) first, but work can begin now with this partial version.

Definition of done:

- [ ] gherkin test suite complete and runs on CI
- [ ] updates to documentation (CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by jamescarr**
## This PR

- adds support to run tests in https://github.com/open-feature/test-harness

### Related Issues

Fixes #99

### CODE REVIEW

**Comment by jamescarr** [comment]
This simply provides the scaffolding: 

- `test-harness` submodule
- python step definitions that raise `NotImplemented`
- Makefile target to run them: `make e2e`
image

File: .gitignore [hunk 1]
```diff
@@ -50,4 +50,6 @@ coverage.xml
 docs/_build/
 
 # Virtual env directories
-.venv
\ No newline at end of file
+.venv
+tests/features/caching.feature
```
Comment (jamescarr) [inline .gitignore:54]: Oops this should be `*.feature`

**Comment by federicobond** [comment]
@jamescarr do you mind if I take a look at this and see if I can get it to work with the new InMemoryProvider?

**Comment by federicobond** [comment]
Rebased against latest main after checking in with @jamescarr and managed to get the evaluation tests to pass with an InMemoryProvider.

I had to fix a couple of minor issues with the design of the provider and also found a bug in the typechecking for python<3.10.

File: Makefile [hunk 2]
```diff
@@ -28,3 +28,10 @@ lint: .venv
 clean:
 	@rm -rf .venv
 	@find -iname "*.pyc" -delete
+
+.PHONY: e2e
+e2e: .venv
+	# NOTE: only the evaluation feature is run for now
+	cp test-harness/features/evaluation.feature tests/features/
+	behave tests/features/
+	rm tests/features/*.feature
```
Comment (toddbaert) [inline Makefile:37]: You may want to add a target that's a prereq to this one that pulls the submodule.

I know the CI does this, and you should keep that, but it's "nice" to have to have it done automatically to make it easy for devs running it locally (people tend to have trouble with git sub-modules in my experience :sweat_smile: ).

File: open_feature/open_feature_client.py [hunk 3]
```diff
@@ -385,7 +385,7 @@ def _typecheck_flag_value(value, flag_type):
     type_map = {
         FlagType.BOOLEAN: bool,
         FlagType.STRING: str,
-        FlagType.OBJECT: typing.Union[dict, list],
+        FlagType.OBJECT: (dict, list),
```
Comment (toddbaert) [inline open_feature/open_feature_client.py:388]: This seems unrelated to the e2e changes. Is it breaking in any way?

**Review by toddbaert — APPROVED** [review]
Largely LGTM, I have a couple minor comments/questions.

I changed the title to "chore" since this doesn't appear to be an "end-user"-facing feature... it's more of a feature for us as OpenFeature devs.

Thanks to both if you for your efforts. Much appreciated as always :pray: @federicobond @jamescarr

File: open_feature/open_feature_client.py (hunk 3 above)
Comment (federicobond) [reply open_feature/open_feature_client.py:388]: Yes, tests did not pass for certain python versions without it. I can extract it to a separate pull request if you want.

**Review by federicobond — APPROVED** [review]
